### PR TITLE
Compat: Limit the number of texture+sampler combinations

### LINKED ISSUE #4986: Compat: Limit the number of sample/texture combinations
In core you can bind 16 textures and 16 samplers (or more). You can use every sampler with every texture (256 combinations). GL can't do that many as every texture unit is tied to specific sampler.

In other words

```wgsl
@group(0) @binding( 0) var t00: texture_2d<f32>;
@group(0) @binding( 1) var t01: texture_2d<f32>;
@group(0) @binding( 2) var t02: texture_2d<f32>;
...
@group(0) @binding(15) var t15: texture_2d<f32>;

@group(1) @binding( 0) var s00: sampler;
@group(1) @binding( 1) var s01: sampler;
@group(1) @binding( 2) var s02: sampler;
...
@group(1) @binding(15) var s15: sampler;

...
   c += textureSample(t00, s00, uv);  
   c += textureSample(t00, s01, uv);  
   c += textureSample(t00, s02, uv);  
   ...
   c += textureSample(t00, s15, uv);  

   ...

   c += textureSample(t01, s00, uv);  
   c += textureSample(t01, s01, uv);  
   c += textureSample(t01, s02, uv);  
   ...
   c += textureSample(t01, s15, uv);  

   ...

   c += textureSample(t02, s00, uv);  
   c += textureSample(t02, s01, uv);  
   c += textureSample(t02, s02, uv);  
   ...
   c += textureSample(t02, s15, uv);  

   ...
   ...

   c += textureSample(t15, s00, uv);  
   c += textureSample(t15, s01, uv);  
   c += textureSample(t15, s02, uv);  
   ...
   c += textureSample(t15, s15, uv);  
```

[example](https://jsgist.org/?src=d6065e0cabd4d4a5f47aac163c7c4768)

Off the top of my head, rather than add some new limit, maybe just state that the number of combinations of samplers and textures can't be > `min(maxSampledTexturesPerShaderStage, maxSamplerPerShaderStage)` in compat?

I'm not sure if that would have any implications for arrays of textures and samplers (https://github.com/gpuweb/gpuweb/issues/822) since it might be hard to check the combinations assuming you can access textures with non-const values. Though I think that restriction exists in GLSL ES 3.1 - "Arrays of samplers may only be indexed by constant-integral-expressions."

### PR DESCRIPTION
**Pull request by greggman**
Resolves #4986

### CODE REVIEW

File: proposals/compatibility-mode.md [hunk 1]
```diff
@@ -209,6 +209,13 @@ In addition to the new limits, the existing `maxStorageBuffersPerShaderStage` an
 
 **Justification**: OpenGL ES 3.1 allows `MAX_VERTEX_SHADER_STORAGE_BLOCKS` and `MAX_VERTEX_IMAGE_UNIFORMS` to be zero, and there are a significant number of devices in the field with that value.
 
+## 20. Limit the number of texture+sampler combinations in a pipeline.
+
+If the number of texture+sampler combinations used a in single stage in a pipeline exceeds
```
Comment (teoxoy) [inline proposals/compatibility-mode.md:214]: Should we clarify that combinations includes textures without samplers as well?

Comment (greggman) [reply proposals/compatibility-mode.md:214]: Yikes. That's an interesting issue.  A texture used without a sampler counts as 1 but only if it's not used somewhere else? Or do we keep it simple? In other words, if you have

```wgsl
@group(0) @binding(0) var t: texture_2d<f32>;
@group(0) @binding(1) var s: sampler;

   x = textureLoad(t);
   y = textureSampler(t, s);
```
   
That's seems like it only counts as 1 combo

I'm not sure the easiest way to state that.

> The number of texture+sampler combinations is counted by first counting the number of texture+sampler combinations, then adding 1 for every texture used without a sampler that's not already counted.

or something like that?

Or do we count the example above as 2 combos? `[(t+s), (t)]`

Comment (greggman) [reply proposals/compatibility-mode.md:214]: Checking dawn with the code above it only created 1 `uniform sampler2D`

Comment (kainino0x) [reply proposals/compatibility-mode.md:214]: I think it should be 1 combo. Probably can be stated as essentially `max(used without sampler ? 1 : 0, number of texture sampler combos)`

Comment (greggman) [reply proposals/compatibility-mode.md:214]: updated

Comment (greggman) [reply proposals/compatibility-mode.md:214]: Is this also complicated by external texture? It seems like it is like maybe it should be `min(maxSampledTexturesPerShaderStage - numExternalTexturesUsedInStage * 4, maxSamplerPerShaderStage - numExternalTexturesUsageInStage * 2)`

Comment (greggman) [reply proposals/compatibility-mode.md:214]: I'm not sure `max(used without sampler ? 1 : 0, number of texture sampler combos)` is correct. I reverted it back to how it was before for now 

`max(used without sampler ? 1 : 0, number of texture sampler combos)` would make sense if it said something like, 

> The number of texture+sampler combinations is counted as follows:
>
> ```
> sum = 0;
> for each texture used in entry point
>    sum += max(texture+sampler combos for texture, texture used without sampler ? 1 :  0)

or something like that

Comment (kainino0x) [reply proposals/compatibility-mode.md:214]: I think that is what I had in mind, I didn't sketch it very carefully. Though just realizing, the 0/1 is accounted for by the bind group layouts, and it doesn't matter if it's actually used with `textureLoad`? So:
```
for each stage of the pipeline:
  sum = 0
  for each texture binding in the pipeline layout which is visible to that stage:
    sum += max(1, number of texture sampler combos for that texture binding)
```

Good point about external texture, that definitely needs to be explicit about how it interacts with this limit. The internal resources of an external texture are explained in the note here: https://gpuweb.github.io/gpuweb/#gpuexternaltexture

So the accounting (https://gpuweb.github.io/gpuweb/#exceeds-the-binding-slot-limits) would add something like this, via the logic above:
```
    sum += 1 // for LUT texture + LUT sampler
    sum += 3 * max(1, number of external_texture sampler combos) // for Y+U+V
```

Comment (greggman) [reply proposals/compatibility-mode.md:214]: ok, added, ptal

**Review by teoxoy — APPROVED** [review]

**Pull request merged**